lib/bolt/config.rb

```diff
@@ -210,169 +237,122 @@ def merge_config_data(config_data)
       end
 
       # Expand paths relative to the Boltdir. Any settings that came from the
       # CLI will already be absolute, so the expand will be skipped.
-      if data.key?('modulepath')
+      if @data.key?('modulepath')
         moduledirs = if data['modulepath'].is_a?(String)
                        data['modulepath'].split(File::PATH_SEPARATOR)
                      else
                        data['modulepath']
                      end
-        @modulepath = moduledirs.map do |moduledir|
+        @data['modulepath'] = moduledirs.map do |moduledir|
           File.expand_path(moduledir, @boltdir.path)
         end
       end
 
-      @inventoryfile = File.expand_path(data['inventoryfile'], @boltdir.path) if data.key?('inventoryfile')
-
-      if data.key?('puppetfile')
-        @puppetfile_config = data['puppetfile'].select { |k, _| PUPPETFILE_OPTIONS.include?(k) }
+      %w[hiera-config inventoryfile trusted-external-command].each do |opt|
+        @data[opt] = File.expand_path(@data[opt], @boltdir.path) if @data[opt]
       end
 
-      @hiera_config = File.expand_path(data['hiera-config'], @boltdir.path) if data.key?('hiera-config')
-      @trusted_external = if data.key?('trusted-external-command')
-                            File.expand_path(data['trusted-external-command'], @boltdir.path)
-                          end
+      # Filter hashes to only include valid options
+      @data['apply_settings'] = @data['apply_settings'].slice(*APPLY_SETTINGS.keys)
+      @data['puppetfile'] = @data['puppetfile'].slice(*PUPPETFILE_OPTIONS.keys)
+    end
 
-      if data.key?('apply_settings')
-        @apply_settings = data['apply_settings'].select { |k, _| APPLY_SETTINGS.keys.include?(k) }
-      end
+    private def normalize_log(target)
+      return target if target == 'console'
+      target = target[5..-1] if target.start_with?('file:')
+      'file:' + File.expand_path(target, @boltdir.path)
+    end
 
-      @compile_concurrency = data['compile-concurrency'] if data.key?('compile-concurrency')
+    private def update_logs(logs)
+      logs.each_with_object({}) do |(key, val), acc|
+        next unless val.is_a?(Hash)
 
-      @save_rerun = data['save-rerun'] if data.key?('save-rerun')
+        name = normalize_log(key)
+        acc[name] = val.slice(*LOG_OPTIONS.keys)
+                       .transform_keys(&:to_sym)
 
-      %w[concurrency format puppetdb color plugins plugin_hooks].each do |key|
-        send("#{key}=", data[key]) if data.key?(key)
-      end
+        if (v = acc[name][:level])
+          unless Bolt::Logger.valid_level?(v)
+            raise Bolt::ValidationError,
+                  "level of log #{name} must be one of #{Bolt::Logger.levels.join(', ')}; received #{v}"
+          end
+          unless v.is_a?(String) || v.is_a?(Symbol)
+            raise Bolt::ValidationError,
+                  "level of log #{name} must be a String or Symbol, received #{v.class} #{v.inspect}"
+          end
```

Comment:
TODO: Remove - It looks like this is redundant, since `#valid_level?` will already return false if the level isn't a valid string or symbol.